core/cas-server-core-services/src/test/java/org/apereo/cas/services/AbstractServicesManagerTests.java

```diff
@@ -180,16 +177,91 @@ public void verifyServiceCanBeFoundDuringLoadWithoutCacheInvalidation() throws E
+     * is matched, it is added to the cache <em>before</em> its associated environments were evaluated in
+     * {@code validateRegisteredService()}: it ends up in cache even though it doesn't match the current environment.
+     * Subsequently, <em>all</em> requests matching {@code https://*.test.edu} will return {@code null} because they are
+     * validated against {@code https://*.test.edu}, which is not associated with {@code current-env}.
+     *
+     * This test verifies that the above situation does not occur. Requests for {@code https://prod.test.edu} should
+     * return a valid service regardless of whether a request for {@code https://[anything else].test.edu} comes in
+     * first. A service which is not associated with the current environment should not interfere.
+     */
+    @Test
+    public void verifyInvalidServiceNotPromotedByCache() {
+        val applicationContext = new StaticApplicationContext();
+        applicationContext.refresh();
+        val cache = Caffeine.newBuilder().expireAfterWrite(Duration.ofDays(1000L)).<Long, RegisteredService>build();
+        val config = ServicesManagerConfigurationContext.builder()
+                .serviceRegistry(serviceRegistry)
+                .applicationContext(applicationContext)
+                .environments(Collections.singleton("current-env"))
+                .registeredServiceLocators(Collections.singletonList(new DefaultServicesManagerRegisteredServiceLocator()))
+                .servicesCache(cache)
+                .build();
+        val manager = new DefaultServicesManager(config);
+
+        /* Service "https://prod.test.edu", low evaluationOrder (high priority), active in "current-env". */
+        val prodTestEduService = new RegexRegisteredService();
+        prodTestEduService.setServiceId("https://prod\\.test\\.edu.*");
+        prodTestEduService.setId(RandomUtils.nextLong());
+        prodTestEduService.setName("Specific Subdomain Service");
+        prodTestEduService.setEvaluationOrder(0);
+        prodTestEduService.setEnvironments(new HashSet<>(Collections.singleton("current-env")));
+
+        /* Service "https://*.test.edu", high evaluationOrder (low priority), NOT active in "current-env". */
+        val anyTestEduService = new RegexRegisteredService();
+        anyTestEduService.setServiceId("https://[^.]+\\.test\\.edu.*");
+        anyTestEduService.setId(RandomUtils.nextLong());
+        anyTestEduService.setName("Any Subdomain Service");
+        anyTestEduService.setEvaluationOrder(1000);
+        anyTestEduService.setEnvironments(new HashSet<>(Collections.singleton("non-current-env")));
+
+        /* Initialize the service manager. */
+        serviceRegistry.save(prodTestEduService);
+        serviceRegistry.save(anyTestEduService);
+        manager.load();
+
+        /* This case always worked: request for https://prod.test.edu before any request for
```

Comment:
Reformat comment and use multi-line.

Reply:
I assume this means *do not* use a multiline comment? Meaning use multiple single-line (//) comments rather than one multiline (/* ... */) comment?

Reply:
You should follow the pattern that is used consistently elsewhere:

```
/* This
is not
good */
```

Instead:

```
/* 
  * This
  * is not
  * good.
  */
```

Something to that effect.

Reply:
Ah, I see. I didn't see any inline comments in the files I was working in, so just used our conventions out of force of habit. I'll update.

Reply:
We generally frown on using inline comments, and prefer javadocs instead. In this case, I think an exception is warranted.

Reply:
Got it, and understood: inline comments tend to drift out of sync with the code over time. This one was a bit hard to explain without chopping it down into steps. Thanks!